src/theory/quantifiers/sygus/sygus_module.h

```diff
@@ -0,0 +1,109 @@
+/*********************                                                        */
+/*! \file sygus_module.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Andrew Reynolds
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2017 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief sygus_module
+ **/
+
+#include "cvc4_private.h"
+
+#ifndef __CVC4__THEORY__QUANTIFIERS__SYGUS_MODULE_H
+#define __CVC4__THEORY__QUANTIFIERS__SYGUS_MODULE_H
+
+#include <map>
+#include "expr/node.h"
+#include "theory/quantifiers_engine.h"
+
+namespace CVC4 {
+namespace theory {
+namespace quantifiers {
+
+class CegConjecture;
+
+/** SygusModule
+ *
+ * This is the base class of sygus modules, owned by CegConjecture.
+ *
+ * An instance of the conjecture class (CegConjecture) creates the negated deep
+ * embedding form of the synthesis conjecture, which assume is:
+ *   forall d. exists x. P( d, x )
+ * where d are of sygus datatype type. The "base instantiation" of this
+ * conjecture (see CegConjecture::d_base_inst) is the formula:
+ *   exists y. P( k, y )
+ * where k are the "candidate" variables for the conjecture.
+ *
+ * Modules implement an initialize function, which determines whether the module
+ * will take responsibility for the given conjecture.
+ */
+class SygusModule
+{
+ public:
+  SygusModule(QuantifiersEngine* qe, CegConjecture* p);
+  ~SygusModule() {}
+  /** initialize
+   *
+   * n is the "base instantiation" of the deep-embedding version of the
+   * synthesis conjecture under candidates (see CegConjecture::d_base_inst).
+   *
+   * This function returns true if this module will take responsibility for
+   * constructing candidates for the given conjecture.
+   */
+  virtual bool initialize(Node n,
+                          const std::vector<Node>& candidates,
+                          std::vector<Node>& lemmas) = 0;
+  /** get term list
+   *
+   * This gets the list of terms that will appear as arguments to a subsequent
+   * call to constructCandidates.
+   */
+  virtual void getTermList(const std::vector<Node>& candidates,
+                           std::vector<Node>& terms) = 0;
+  /** construct candidate
+   *
+   * terms : the terms returned by a call to getTermList,
+   * term_values : the current model values of terms,
+   * candidates : the list of candidates.
+   *
+   * If this function returns true, it adds to candidate_values a list of terms
+   * of the same length and type as candidates that are candidate solutions
+   * to the synthesis conjecture in question. This candidate { v } will then be
+   * tested by testing the (un)satisfiablity of P( v, k' ) for fresh k'.
```

Comment:
Clarify in which conditions this function would return true. Explain also in which context this function is called. And it seems that this is the function testing the candidate values, which I don't think is the case?

Reply:
candidate_values is empty when this function is called, the job of this function is to add a candidate-to-test to candidate_values.